### ISSUE TITLE
[event][filter_attack]special  also triggers on inactive specials

### ISSUE THREAD

**Issue by gfgtdf** [wesnoth/wesnoth#3317]
as said in the title, [event][filter_attack]special=<id> also passes if the speial is inactive, which makes it harder to implement specials with a custom filter. testing code:

```
local u = wesnoth.get_units({side = 1})[1]

u:add_modification("object", {
	T.effect {
		apply_to ="attack",
		T.set_specials {
			T.drains {
				id = "g_test",
				T.filter_self {
					id ="asdfasdfsa"
				}
			}
		}
	}
})

wesnoth.wml_actions.event {
	name="attack_end",
	T.filter_attack{
		special = "g_test",
	},
	T.message {
		message = "aaaa"
	}
}
```

the message gets printed when the unit attacks, even though the unit id is not "asdfasdfsa" and the filter_self does not pass (the unit also does not drain)

As this won't get fixed, i am also interested to know aout any other existing lua/wml method to tell whether an attack special is currently active or not.

**Comment by Vultraz** [comment]
Why do you say it won't get fixed? This is a perfectly reasonable thing to fix, or at least add another key to enable.

**Comment by CelticMinstrel** [comment]
In my opinion this should be fixed if at all possible, as that's kinda part of the purpose of SUF `special=`.

**Comment by CelticMinstrel** [comment]
In my opinion this should be fixed if at all possible, as that's kinda part of the purpose of SUF `special=`.

**Comment by gfgtdf** [comment]
> Why do you say it won't get fixed?

I was assuming that it's somehow an apichange that's unsuitable for 1.14. I wouldn't object if someone did it though in particular is it happens before the next 1.14 release

**Comment by CelticMinstrel** [comment]
Well, it definitely needs to be fixed. I'll let someone else decide whether 1.14 can get the fix (someone who knows whether it would create compatibility issues between 1.14 versions).

**Comment by gfgtdf** [comment]
ok it read the suorce ode and it seems like 'special_active 'is the attrubte that should be used there but even that part doesn't seems to work, when i use `special_active` it just never triggers, new testing  code:
```
u:add_modification("object", {
	T.effect {
		apply_to ="attack",
		T.set_specials {
			T.drains {
				id = "g_test",
				T.filter_self {
					side = 1
				}
			}
		}
	}
})

wesnoth.wml_actions.event {
	name="attack_end",
	T.filter_attack{
		special_active = "g_test",
	},
	T.message {
		message = "aaaa"
	}
}
```

not as the unit is of side 1 the filter should alway pass as the unit is of side 1, the draining also happens but the event never fires.


ps: when this gets fixed we shoudl also update https://wiki.wesnoth.org/EventWML#.5Bfilter_attack.5D to suggest using special_active  over special

**Comment by CelticMinstrel** [comment]
I think this just means that some places need to set the weapon specials context. It might also turn out that there are places where `special_active` will just never work though, like in an object filter.

**Comment by gfgtdf** [comment]
> I think this just means that some places need to set the weapon specials context

well to be more precise the problme is that in [event] [filter_attack] filter is not applied the the actual attack of that unit (which does have the context set), but to a copy of that attack for some resaon. in particular this gives me a way to workaround this in my addons: apply the filter manually via lua to the attack of the unit instead (fof that i'd have to get the attack object though which i can only do if i assume that attack id to be unique, which might be reasonable in my case),

**Comment by CelticMinstrel** [comment]
I see, that kinda makes sense; but in theory it should still be possible to set the weapon specials context on that copy of the attack, shouldn't it?

**Comment by ln-zookeeper** [comment]
Yes, in an attack filter special= (just like ability= in a SUF) should **not** be limited to matching only active specials. Since special_active= doesn't make sense in attack filters in general, it could just be an extra key allowed in [filter_attack] and similar tags.

**Comment by gfgtdf** [comment]
> Yes, in an attack filter special= (just like ability= in a SUF) should not be limited to matching only active specials

well yes, it makes sense as there is specials_active for that, but im curious what possible usecases of the `special=` attribute could be.

> Since special_active= doesn't make sense in attack filters in general, it could just be an extra key allowed in [filter_attack] and similar tags.

It can easily be well defined: jJust define it as false whenever the units attack  is not currently involved in a attack.

More importantly your suggestion souns like it would break the code
```
wesnoth.wml_actions.event {
	name="attack_end",
	T.filter {
		T.has_attack {
			special_active = "g_test",
		}
	},
	T.message {
		message = "aaaa"
	}
}
```
which (unlike my code examples above) does work on 1.14 if you you special has a [filter_self] and only triggers the mesage of the units currently selected attack has an spacial active with that id.

**Comment by CelticMinstrel** [comment]
Um, what? The `special=` key is to match a weapon with a specific special. For example, if you want to add an object that applied an effect to all magical attacks.

**Comment by ProditorMagnus** [comment]
I ran into this with https://github.com/ProditorMagnus/Ageless-for-1-14/issues/18

Do I understand correctly that special_active only makes sense as check for:
- [filter_adjacent]
- [filter_adjacent_location]
- [filter_self] without $other_unit

And any of following means that special is never active:
- active_on
- apply_to=opponent/attacker/defender
- any filters depending on attacker/defender

And the following do not invalidate match:
- apply_to=self/both

**Comment by gfgtdf** [comment]
> Do I understand correctly that special_active only makes sense as check for:

do you read the somehow from this conversation or is it something that you found out?

(honestly i don't really remember this topic, all i remmbr is that i ended up adding empty [filter_self][/filter_self] to by addon becasue otherwise spacil_active does not work.)

**Comment by gfgtdf** [comment]
_I think_ not sure, that is also better to use filter_unit with has_attack special_active over filter_attack, and also make sure each of the specials that are filtered have a (potentially empty) [filter_self] tag.

**Comment by ProditorMagnus** [comment]
I tested with apply_to=self/opponent, and active_on=offense. Others are just my guess based on documentation and how I think special_active could have been implemented.

**Comment by gfgtdf** [comment]
sry i don't really understand, can you give me an example of a code that does not work?

**Comment by ProditorMagnus** [comment]
I used charge as check for active_on.

```
    [event]
        name=attack
        first_time_only=no
        [filter_attack]
            special_active=charge
        [/filter_attack]
		[lua]
			code=<<wesnoth.message("charge active")>>
		[/lua]
    [/event]

    [event]
        name=attack
        first_time_only=no
        [filter_attack]
            special=charge
        [/filter_attack]
		[lua]
			code=<<wesnoth.message("charge exists")>>
		[/lua]
    [/event]
```

**Comment by gfgtdf** [comment]
> I used charge as check for active_on.

okay but you are using [filter_attack], what happens when you use [has_attack] instead?
```
    [event]
        name=attack
        first_time_only=no
        [filter]
[has_attack]]
            special_active=charge
[/has_attack]]
        [/filter]
		[lua]
			code=<<wesnoth.message("charge active")>>
		[/lua]
    [/event]

    [event]
        name=attack
        first_time_only=no
        [filter_attack]
            special=charge
        [/filter_attack]
		[lua]
			code=<<wesnoth.message("charge exists")>>
		[/lua]
    [/event]

```

**Comment by gfgtdf** [comment]
wait, wh are you using charge in the first place? it iirc has no filters so there is no reason to use special_active?

**Comment by ProditorMagnus** [comment]
Charge has filter - it is only active when attacking.

**Comment by ProditorMagnus** [comment]
I tested that there is difference when defending - unit with charge defends against unit with charge.
Results in "has second charge active" without "second charge active".

```
	[event]
        name=attack
        first_time_only=no
		[filter_second]
			[has_attack]
				special_active=charge
			[/has_attack]
		[/filter_second]
		[lua]
			code=<<wesnoth.message("has second charge active")>>
		[/lua]
    [/event]

    [event]
        name=attack
        first_time_only=no
        [filter_second_attack]
            special_active=charge
        [/filter_second_attack]
		[lua]
			code=<<wesnoth.message("second charge active")>>
		[/lua]
    [/event]
```

**Comment by ProditorMagnus** [comment]
After I removed apply_to=both from {WEAPON_SPECIAL_CHARGE} "has second charge active" was gone too. So it looks like apply_to=both internally adds weapon special (temporarily) to the other unit.

**Comment by ProditorMagnus** [comment]
I believe this issue may be fixed by now. At least the same test results are no longer reliable and need retesting.
